crates/nu-plugin/src/plugin/context.rs

```diff
@@ -1,36 +1,161 @@
-            ctrlc: engine_state.ctrlc.clone(),
+            identity,
+            engine_state: engine_state.clone(),
+            stack: stack.clone(),
+            call: call.clone(),
         }
     }
 }
 
 impl PluginExecutionContext for PluginExecutionCommandContext {
+    fn command_span(&self) -> Span {
+        self.call.head
+    }
+
+    fn command_name(&self) -> &str {
+        self.engine_state.get_decl(self.call.decl_id).name()
+    }
+
     fn ctrlc(&self) -> Option<&Arc<AtomicBool>> {
-        self.ctrlc.as_ref()
+        self.engine_state.ctrlc.as_ref()
+    }
+
+    fn get_config(&self) -> Result<Config, ShellError> {
+        Ok(nu_engine::get_config(&self.engine_state, &self.stack))
+    }
+
+    fn get_plugin_config(&self) -> Result<Option<Value>, ShellError> {
+        // Fetch the configuration for a plugin
+        //
+        // The `plugin` must match the registered name of a plugin.  For
+        // `register nu_plugin_example` the plugin config lookup uses `"example"`
+        Ok(self
+            .get_config()?
+            .plugins
+            .get(&self.identity.plugin_name)
+            .cloned()
+            .map(|value| {
+                let span = value.span();
+                match value {
+                    Value::Closure { val, .. } => {
+                        let input = PipelineData::Empty;
+
+                        let block = self.engine_state.get_block(val.block_id).clone();
+                        let mut stack = self.stack.captures_to_stack(val.captures);
+
+                        match eval_block_with_early_return(
+                            &self.engine_state,
+                            &mut stack,
+                            &block,
+                            input,
+                            false,
+                            false,
+                        ) {
+                            Ok(v) => v.into_value(span),
+                            Err(e) => Value::error(e, self.call.head),
+                        }
+                    }
+                    _ => value.clone(),
+                }
+            }))
+    }
+
+    fn eval_closure(
+        &self,
+        closure: Spanned<Closure>,
+        positional: Vec<Value>,
+        input: PipelineData,
+        redirect_stdout: bool,
+        redirect_stderr: bool,
+    ) -> Result<PipelineData, ShellError> {
+        let block = self
+            .engine_state
+            .try_get_block(closure.item.block_id)
+            .ok_or_else(|| ShellError::GenericError {
+                error: "Plugin misbehaving".into(),
+                msg: format!(
+                    "Tried to evaluate unknown block id: {}",
+                    closure.item.block_id
+                ),
+                span: Some(closure.span),
+                help: None,
+                inner: vec![],
+            })?;
+
+        let mut stack = self.stack.captures_to_stack(closure.item.captures);
+
+        // Set up the positional arguments
```

Comment:
I believe this should be more complicated.  e.g: What if a closure contains optional parameters like: `{|indx: int = 3, ...rest| $indx + 1}`.  And I believe it can be handled by the same logic which introduce by this pr: https://github.com/nushell/nushell/pull/12056

Can you add a `TODO` comment to handle these cases?  I can try to help on it once #12056 is ok.

Reply:
I don't think that has any impact on this - because `eval_call` is eventually called anyway, and all of the validation happens there, shouldn't it just work? You can pass any number of args you want back to the engine, and the engine will handle putting them in the right place for the closure - even if there are optional or rest args

Reply:
I think the best resolution for this would be to add tests for these, and then add some kind of special logic if the tests don't pass - based on reading your PR, I think it should work fine, but we can always check.

Reply:
> because eval_call is eventually called anyway, and all of the validation happens there, shouldn't it just work?

Actually I don't think `eval_call` will be called, because ` eval_block_with_early_return` won't call `eval_call` for the same block.

-----
But I agree with you, we don't need to add such logic here.  Because closure argument type is already defined in signature, there is no case that user passes a closure that have optional parameter(or it's useless)

Reply:
I will still add tests anyway.